sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -251,6 +262,77 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
     }
   }
 
+  // Get the bucket ID based on the bucketing values.
+  // Restriction: Bucket pruning works iff the bucketing column has one and only one column.
+  private def getBucketId(bucketColumn: Attribute, numBuckets: Int, value: Any): Int = {
+    lazy val mutableRow = new SpecificMutableRow(Seq(bucketColumn.dataType))
```

Comment:
This won't reuse the row. Anyway it's not in critical path, we can leave it as simple `val`